graphics/tinygl/light.cpp

```diff
@@ -228,88 +230,93 @@ void gl_shade_vertex(GLContext *c, GLVertex *v) {
 			d.X = l->position.X - v->ec.X;
 			d.Y = l->position.Y - v->ec.Y;
 			d.Z = l->position.Z - v->ec.Z;
 			dist = sqrt(d.X * d.X + d.Y * d.Y + d.Z * d.Z);
-			if (dist > 1E-3) {
-				tmp = 1 / dist;
-				d *= tmp;
-			}
 			att = 1.0f / (l->attenuation[0] + dist * (l->attenuation[1] +
 			              dist * l->attenuation[2]));
 		}
 		dot = d.X * n.X + d.Y * n.Y + d.Z * n.Z;
 		if (twoside && dot < 0)
 			dot = -dot;
 		if (dot > 0) {
+			if (dist != 1) {
+				tmp = 1 / dist;
+				d *= tmp;
+				dot *= tmp;
+			}
 			// diffuse light
 			lR += dot * l->diffuse.X * m->diffuse.X;
 			lG += dot * l->diffuse.Y * m->diffuse.Y;
 			lB += dot * l->diffuse.Z * m->diffuse.Z;
 
-			// spot light
-			if (l->spot_cutoff != 180) {
-				dot_spot = -(d.X * l->norm_spot_direction.X +
-							 d.Y * l->norm_spot_direction.Y +
-							 d.Z * l->norm_spot_direction.Z);
-				if (twoside && dot_spot < 0)
-					dot_spot = -dot_spot;
-				if (dot_spot < l->cos_spot_cutoff) {
-					// no contribution
-					continue;
-				} else {
-					// TODO: optimize
-					if (l->spot_exponent > 0) {
-						att = att * pow(dot_spot, l->spot_exponent);
+			const bool is_spotlight = l->spot_cutoff != 180;
+			const bool has_specular = l->has_specular && m->has_specular;
+			if (is_spotlight || has_specular) {
+				if (is_spotlight) {
+					dot_spot = -(d.X * l->norm_spot_direction.X +
+								 d.Y * l->norm_spot_direction.Y +
+								 d.Z * l->norm_spot_direction.Z);
+					if (twoside && dot_spot < 0)
+						dot_spot = -dot_spot;
+					if (dot_spot < l->cos_spot_cutoff) {
+						// no contribution
+						continue;
+					} else {
+						// TODO: optimize
+						if (l->spot_exponent > 0) {
+							att = att * pow(dot_spot, l->spot_exponent);
+						}
 					}
 				}
-			}
 
-			// specular light
-
-			if (c->local_light_model) {
-				Vector3 vcoord;
-				vcoord.X = v->ec.X;
-				vcoord.Y = v->ec.Y;
-				vcoord.Z = v->ec.Z;
-				vcoord.normalize();
-				s.X = d.X - vcoord.X;
-				s.Y = d.Y - vcoord.Y;
-				s.Z = d.Z - vcoord.Z;
-			} else {
-				s.X = d.X;
-				s.Y = d.Y;
-				s.Z = (float)(d.Z + 1.0);
-			}
-			dot_spec = n.X * s.X + n.Y * s.Y + n.Z * s.Z;
-			if (twoside && dot_spec < 0)
-				dot_spec = -dot_spec;
-			if (dot_spec > 0) {
-				GLSpecBuf *specbuf;
-				int idx;
-				tmp = sqrt(s.X * s.X + s.Y * s.Y + s.Z * s.Z);
-				if (tmp > 1E-3) {
-					dot_spec = dot_spec / tmp;
+				if (has_specular) {
+					if (c->local_light_model) {
+						Vector3 vcoord;
+						vcoord.X = v->ec.X;
+						vcoord.Y = v->ec.Y;
+						vcoord.Z = v->ec.Z;
+						vcoord.normalize();
+						s.X = d.X - vcoord.X;
+						s.Y = d.Y - vcoord.Y;
+						s.Z = d.Z - vcoord.Z;
+					} else {
+						s.X = d.X;
+						s.Y = d.Y;
+						s.Z = (float)(d.Z + 1.0);
+					}
+					dot_spec = n.X * s.X + n.Y * s.Y + n.Z * s.Z;
+					if (twoside && dot_spec < 0)
+						dot_spec = -dot_spec;
+					if (dot_spec > 0) {
+						GLSpecBuf *specbuf;
+						int idx;
+						tmp = sqrt(s.X * s.X + s.Y * s.Y + s.Z * s.Z);
```

Comment:
Second comment no longer applies, but I believe there is still an opportunity to avoid the expensive sqrt call. I just don't know how to get the compiler to emit a reciprocal square root without intrinsics...